design/security/high_availability.md

```diff
@@ -0,0 +1,66 @@
+# High availability
+
+The Obscuro Sequencer must have HA capabilities. The reasoning is covered in the "Fast finality" design docs.
+
+The requirement is that the service must continue even if the enclave of the sequencer crashes and is corrupted.
+
+One option is to require a governance event to whitelist a new sequencer enclave in case the current one becomes unavailable.
+This is not a good solution, because it would lead to significant downtime.
+
+The preferred option is to mark multiple enclaves as sequencers, and make it the responsibility of the operator to switch
+between them to ensure that service goes on uninterrupted.
+
+## MEV Protection 
+
+The Sequencer is the only Obscuro node capable of producing Light Batches, and thus the only node that can in theory
+attempt to extract value via MEV. If the operator has multiple enclaves available it can both provide an uninterrupted service
+and extract some value at the same time.
+
+Introducing startup delays does not help too much in this case, because the operator could hold key transactions for longer.
+
+
+An alternative solution is to introduce restart transparency, such that the Obscuro network can assess the likelyhood
+of bad behaviour.  
+
+On a high level, all enclave restart events of all the sequencer whitelisted enclaves must be published to the network.
+
+### Protocol
+
+Each Batch will have three elements:
+
+1. The header
+2. The transaction payload
+3. The protocol payload
+
+The protocol payload will not be included in the rollup published to the data availability layer. 
+
+Note: the protocol payload can be used for other protocol specific messages, like current attestations.
+
+```golang
+type RestartEvent struct {
```

Comment:
I think we need a unique ID for each restart - basically, get a random number and stick it. The lastRestartTime might be easy to manipulate and the rest _might_ be manipulatable to some extent. The unique ID will always identify different runs of the enclave so seeing a lot of rollups with different IDs would be a cause for alarm. I guess one can even monitor them for 99.999% uptime or whatever.

Reply:
That's roughly why I included the `batchHeadAtRestart` and the `currentBatchHead`.

The `batchHeadAtRestart` - is pretty much the equivalent of the nonce, I'd say. 
The point of the  `currentBatchHead` - is to prevent the host from reusing events.

Thinking about this, the host could restart Enclave 2 multiple times, and not include any event, and then after a couple of minutes, restart it from a previous point in time and claim it was up all this time, and there was just one restart.

This mechanism has to be somehow connected with the replay protection from the other document.
 
Back to the drawing board ..

Reply:
> Thinking about this, the host could restart Enclave 2 multiple times, and not include any event, and then after a couple of minutes, restart it from a previous point in time and claim it was up all this time, and there was just one restart.

When we talked about this in the past I think we said the enclave won't start functioning until it sees that its restart has been broadcast somehow. Maybe even via the L1? Note: this doesn't slow down the failover process, it slows down the time for the previous sequencer to be re-included in the HA pool.